### ISSUE TITLE
Mock's don't seem to honor sequential ordering

### ISSUE THREAD

**Issue by yaronyg** [sinonjs/sinon#983]
I would expect the following test to fail:

```javascript
  var foo = { m: function () {}};
  var mock = sinon.mock(foo);
  mock.expects('m').once().withExactArgs(23).returns(23);
  mock.expects('m').once().withExactArgs(45).returns(45);
  console.log(foo.m(45));
  console.log(foo.m(23));
  mock.verify();
```
Because once was set first for 23 and then for 45. But in fact, the test passes even though 45 was called before 23. My expectation that this would work at all was based on posts like [this one](https://groups.google.com/forum/#!topic/sinonjs/QSx5a-ar38k).

I'm trying to test that the mock is called in a specific order with specific arguments. Should the above have worked and this is a bug? Is the above bad but there is a right way? Am I just out of luck without doing after the fact validation using spies? Any help is appreciated!

**Comment by fatso83** [comment]
I am not familiar with the mocks functionality, and since it is going away in Sinon 2.0 I don't see myself becoming more familiar with it either ;-) This might be answerable by someone with more intimate with mocks. I cannot do it by glancing over the code.

**Comment by yaronyg** [comment]
Until 2.0 is around what is the best way to write mock like code that will have some hope of forward compatibility? Should I just use a combination of stubs and spies or are they going away as well?

Is there anything intended to replace mocks? The interface was really amazing for our needs because we are testing an object whose job it is to manage another object. So we mock the dependent object and set the expectations for how it should work in the test. If the object under test doesn't manage the mocked object correctly then we get an instant error. It's also much easier to read the test because we can specify all of our expectations of the mock object, both in terms of what inputs it should get and what outputs it should return in one place. With stubs and spies we have to separate everything out and we can only find problems after the fact when we check the spies.

Any help or pointers is appreciated.

**Comment by yaronyg** [comment]
BTW, I poked around the Sinon GitHub depot and the docs on master (which lists its version as 2.0.0.-pre) still list mock and the mock object is still included. Where do I go to learn about 2.0 then?

**Comment by fatso83** [comment]
Sorry for keeping you waiting. As you can see in #991, documentation for version 2 is on the to-do list. Feel free to contribute :-) #966 is where the work is being done /compiled. 

As for where to learn more about the discussions around mocks, see #745. You can see that I misinformed you: mocks are being deprecated in 2, and will be removed in version 3. But you can still use them! They will just be extracted to a separate package, that's all.

**Comment by yaronyg** [comment]
Fair enough. I think it's unfortunate because Mocks were perfect for our style of testing. We had a situation where state machine A was calling state machine B and B was mocked. What we were testing is that under a given set of stimuli A would call B with the right API calls in the right order. Mocks would have provided us with a really simple, compact way to do this.

Instead what we have had to do is stub B and then set up a fairly complicated series of spies where we can define the exact call ordering. So, for example, A would be calling 'start' and 'stop' APIs on B. The key is that the arguments and ordering of the calls mattered. So we needed an output like "B.start(foo), B.start(bar), B.stop(), B.start(bar), B.stop()".

In theory expressing that as a mock is easy. We obviously can validate this using spies but it's tough as we have to create a bunch of separate spies and play games with 'beforeUse' and 'afterUse'. It just made things much uglier. It also made the test configuration more complex because we configured stubs in one place and then we had to specify the spies in a separate place (that would run later). With mocks we could do everything at once.

But I don't have the cycles to fix mocks so I'll just close this issue as won't fix.
